File: lib/vimade/v2/state/globals.py
import sys
import time
import vim
from vimade.v2.recipe import default as DEFAULT_RECIPE
from vimade.v2.style.value import animate as ANIMATE
from vimade.v2.style.value import condition as CONDITION
from vimade.v2.style.value import direction as DIRECTION
from vimade.v2.style.value import ease as EASE
from vimade.v2.style import fade as FADE
from vimade.v2.style import tint as TINT
from vimade.v2.util import ipc as IPC
from vimade.v2.util import matchers as MATCHERS
from vimade.v2.util import type as TYPE

MAX_TICK_ID = 1000000000

_OTHER = {
  'vimade_fade_active': False,
  'is_dark': False
}
_CURRENT = {
  'winid': -1,
  'winnr': -1,
  'bufnr': -1,
  'tabnr': -1,
}
_DEFAULTS = TYPE.extend(DEFAULT_RECIPE.Default(), {
  'basebg': None,
  'ncmode': 'buffers',
  'tint': None,
  'fadelevel': 0.4,
  'fademinimap': None,
  'groupdiff': True,
  'groupscrollbind': False,
  # link can be an array of objects or functions
  # objects are passed to the default handler
  # values can be Matchers
  # functions are passed (Win, ActiveWin, value)
  'link': {
    'default': {
      'buf_name': None,
      'buf_opts': None,
      'buf_vars': None,
      'win_opts': None,
      'win_vars': None,
      'win_config': None,
    }
  },
  # link can be an array of objects or functions
  # objects are passed to the default handler
  # values can be Matchers
  # functions are passed (Win, ActiveWin, value)
  'blocklist': {
    'default': {
      'buf_name': None,
      'buf_opts': {
        'buftype': ['popup', 'prompt']
      },
      'buf_vars': None,
      'win_opts': None,
      'win_vars': None,
      'win_type': True,
      'win_config': {
        'relative': True #block all floating windows # TODO we can make this more customized soon
       },
    }
  },
  # python only config
  'enabletreesitter': False,
  'enablebasegroups': False,
  'basegroups': [],
  'rowbufsize': 0,
  'colbufsize': 0,
  'enablescroll': True,
  'enablesigns': True,
  'signsid': 13100,
  'signspriority': 31,
  'signsretentionperiod': 4000,
  'disablebatch': False,
})

class Globals(object):
  def __init__(self):
    # python specific
    (is_nvim,
     is_gui_running,
     original_background,
     features) = IPC.eval_and_return('['+','.join([
       'has("nvim")',
       'has("gui_running")',
       '&background',
       'g:vimade_features',
     ])+']')

    self.__internal = {
      'READY': 0,
      'ERROR': 1,
      'CHANGED': 2,
      'INVALIDATE_HIGHLIGHTS': 4,
      'INVALIDATE_BUFFER_CACHE': 8,
      'RECALCULATE': 16,
      'SIGNS': 32,
      'features': features,
      'vimade_py': {},
      'tick_id': 1000,
      'tick_state': 0,
      'vimade_fade_active': False,
      'basebg': None,
      'normalid': 0,
      'normalncid': 0,
      'ncmode': 'buffers',
      'matchpriority': 0,
      'nc_windows': False,
      'nc_buffers': False,
      'tint': None,
      'fadelevel': 0,
      'fademinimap': False,
      'groupdiff': True,
      'groupscrollbind': False,
      'colorscheme': None,
      'is_dark': False,
      'current': {
          'winid': -1,
          'winnr': -1,
          'bufnr': -1,
          'tabnr': -1,
      },
      'link': {},
      'blocklist': {},
      'is_nvim': bool(int(is_nvim)),
      'is_term': bool(int(is_gui_running)) == False,
      'rowbufsize': None,
      'colbufsize': None,
      'enablescroll': False,
      'enablesigns': False,
      'enabletreesitter': False,
      'enablebasegroups': False,
      'basegroups': [],
      'signsretentionperiod': 0,
      'signsid': None,
      'signspriority': None,
      'winwidth': 20,
      'require_treesitter': False,
      'termguicolors': None,
      'disablebatch': False,
      'style': [],
      'now': None,
    }
    self.vim = vim
    self.time = time
    self.FADE = FADE
    self.TINT = TINT
    self.IPC = IPC
    self.MATCHERS = MATCHERS
    self.MAX_TICK_ID = MAX_TICK_ID
    self.TYPE = TYPE
    self._OTHER = _OTHER
    self._CURRENT = _CURRENT
    self._DEFAULTS = _DEFAULTS
    sys.modules[__name__] = self

  def __getattribute__(self, key):
    if key == '_Globals__internal':
      return object.__getattribute__(self, key)
    elif not key in self.__internal:
      return object.__getattribute__(self, key)
    else:
      return object.__getattribute__(self, '_Globals__internal')[key]
  def __setattr__(self, key, value):
    if key == '_Globals__internal':
      object.__setattr__(self, key, value)
    else:
      object.__getattribute__(self, '_Globals__internal')[key] = value
    return value

  def __getitem__(self, name):
    return self.__internal[name]

  def __setitem__(self, name, value):
    self.__internal[name] = value

  def _check_fields(self, fields, next, current, defaults, return_state):
    modified = False
    for field in fields:
      value = next.get(field, defaults.get(field))
      if current[field] != value:
        current[field] = value
        modified = True
    return return_state if modified else self.READY

  def _next_tick_id(self):
    tick_id = self.tick_id + 1
    if tick_id > self.MAX_TICK_ID:
      tick_id = 1000
    return tick_id

  def setup(self, config):
    self.vimade_py = self.TYPE.deep_copy(config)

  def getInfo(self):
    info = self.__internal
    result = {'renderer': 'python-v2'}
    exclude = ['vim', 'time', 'TINT', 'IPC', '_OTHER', '_CURRENT', '_DEFAULTS']
    for name in info.keys():
      if name[:2] != '__' and not name in exclude:
        item = info[name]
        if item == None or isinstance(item, bool) or isinstance(item, int) or isinstance(item, float) or isinstance(item, str) or isinstance(item, dict) or isinstance(item, list):
          result[name] = item
    return result

  def refresh(self, tick_state = 0):
    self.tick_id = self._next_tick_id()
    self.tick_state = tick_state
    self.now = self.time.time()
    (background,
      colorscheme,
      termguicolors,
      winwidth,
      vimade_fade_active,
      winid,
      winnr,
      bufnr,
      tabnr) = self.IPC.mem_safe_eval('['+','.join([
      '&background',
      'exists("g:colors_name") ? g:colors_name : ""',
      '&termguicolors',
      '&winwidth',
      'exists("g:vimade_fade_active") ? g:vimade_fade_active : ""',
      'win_getid()',
      'winnr()',
      'bufnr()',
      'tabpagenr()',
    ])+']')
    # we need the actual converted object here for config. This coercion is
    # required for consistency between nvim and vim.
    # some versions of vim return bytes instead of str
    # output of coerceTypes is essentially already a deep copy
    vimade = self.TYPE.shallow_extend(self.IPC.coerceTypes(self.vim.vars['vimade']), self.vimade_py)
    if 'fadepriority' in vimade and not 'matchpriority' in vimade:
      vimade['matchpriority'] = vimade['fadepriority']
    

    current = {
      'winid': int(winid),
      'winnr': int(winnr),
      'bufnr': int(bufnr),
      'tabnr': int(tabnr)
    }

    self.tick_state |= self._check_fields([
      'is_dark',
      'colorscheme',
      'termguicolors'
    ], {
      'is_dark': background == 'dark', # these settings trigger an initial RECALCULATE on load (this sets things like vimade_0)
      'colorscheme': colorscheme,
      'termguicolors': bool(int(termguicolors))
    }, self, self._OTHER, self.INVALIDATE_HIGHLIGHTS | self.RECALCULATE | self.CHANGED)

    # neovim only - no tricky
    if self.is_nvim:
      if int(vimade.get('enabletreesitter')) and not self.require_treesitter:
        vim.api.exec_lua("_vimade_legacy_treesitter = require('vimade_legacy_treesitter')", [])
        self.require_treesitter = 1
    else:
      vimade['enabletreesitter'] = 0
      vimade['enablebasegroups'] = 0

    # coerce int here due to vim+python long issue
    vimade['normalid'] = int(vimade['normalid'])
    vimade['normalncid'] =  int(vimade['normalncid']) if vimade['normalncid'] != '' else None
    self.tick_state |= self._check_fields([
      'normalid',
      'normalncid',
      'enablebasegroups',
      'matchpriority',
    ], vimade, self, self._DEFAULTS, self.INVALIDATE_HIGHLIGHTS | self.CHANGED)
    self.tick_state |= self._check_fields([
      'enabletreesitter'
    ], vimade, self, self._DEFAULTS, self.INVALIDATE_BUFFER_CACHE | self.INVALIDATE_HIGHLIGHTS | self.CHANGED)
    self.tick_state |= self._check_fields([
      'vimade_fade_active'
    ], {
      'vimade_fade_active': bool(int(vimade_fade_active))
    }, self, self._OTHER, self.CHANGED)
    self.tick_state |= self._check_fields([
      'ncmode',
      'enablescroll',
      'colbufsize',
      'rowbufsize'
    ], vimade, self, self._DEFAULTS, self.CHANGED)
    self.tick_state |= self._check_fields([
      'winid',
      'winnr',
      'bufnr',
      'tabnr',
    ], current, self.current, self._CURRENT, self.READY) #TODO just set
    self.tick_state |= self._check_fields([
      'enablesigns'
    ], vimade, self, self._DEFAULTS, self.SIGNS)

    ## handled in win state
    self.style = vimade.get('style', self._DEFAULTS['style'])
    self.basegroups = vimade.get('basegroups', self._DEFAULTS['basegroups'])
    self.signsid = int(vimade.get('signsid', self._DEFAULTS['signsid']))
    self.signspriority = int(vimade.get('signspriority', self._DEFAULTS['signspriority']))
    self.signsretentionperiod = int(vimade.get('signsretentionperiod', self._DEFAULTS['signsretentionperiod']))
    self.link = vimade.get('link', self._DEFAULTS['link']) # TODO this should be singlekey merge (allow users to override the default or add additional config)
    self.blocklist = vimade.get('blocklist', self._DEFAULTS['blocklist']) # TODO this should be singlekey merge (allow users to override the default or add additional config)
    self.basebg = vimade.get('basebg', self._DEFAULTS['basebg']) # TODO empty string needed? 
    self.groupdiff = bool(int(vimade.get('groupdiff', self._DEFAULTS['groupdiff'])))
    self.groupscrollbind = bool(int(vimade.get('groupscrollbind', self._DEFAULTS['groupscrollbind'])))
    self.fademinimap = bool(int(vimade.get('fademinimap', self._DEFAULTS['fademinimap'])))
    self.tint = vimade.get('tint', self._DEFAULTS['tint'])
    self.fadelevel = float(vimade.get('fadelevel', self._DEFAULTS['fadelevel']))
    self.disablebatch = bool(int(vimade.get('disablebatch', self._DEFAULTS['disablebatch'])))

    ## already checked
    self.nc_windows = self.ncmode == 'windows'
    self.nc_buffers = self.ncmode == 'buffers'
    # if you don't choose one of the above, everything is highlighted

M = Globals()

File: lua/vimade/state/globals.lua
local M = {}

local bit = require('bit')
local BIT_BAND = bit.band
local BIT_BOR = bit.bor

local FOCUS = require('vimade.focus')
local TYPE = require('vimade.util.type')
local REAL_NAMESPACE = require('vimade.state.real_namespace')
local DEFAULT_RECIPE = require('vimade.recipe.default')

-- arbitrary high number for us to regulate what has been processed in a single tick
local MAX_TICK_ID = 400000
local next_tick_id = function ()
  local tick_id = M.tick_id + 1
  if tick_id > MAX_TICK_ID then
    tick_id = 1
  end
  return tick_id
end

M.READY = 0
M.ERROR = 1
M.CHANGED = 2
M.RECALCULATE = 4
M.DISCARD_NS = 8
-- some plugins (e.g. Neotree) add the highlights on window/buffer change. When nohlcheck,
-- which prevents rechecking the namespaces for no reason, is enabled we tell the downstream
-- logic to check namespaces on buf/win/tab change.
M.HLCHECK = 8

M.vimade_lua = {}

M.tick_id = 0
M.tick_state = M.READY
M.now = vim.loop.now()
M.global_ns = nil-- used to cache namespace 0 at the global level
M.vimade_fade_active = false
M.vimade_focus_active = false
M.basebg = nil
M.normalid = 0
M.normalncid = 0
M.ncmode = 'buffers'
M.nc_windows = false
M.nc_buffers = false
M.fadelevel = 0
M.fademinimap = false
M.groupdiff = true
M.groupscrollbind = false
M.nohlcheck = true
M.colorscheme = nil
M.termguicolors = nil
M.is_dark = false
M.current = {
  winid = -1,
  bufnr = -1,
  tabnr = -1,
}
M.link = {}
M.blocklist = {}
M.global_highlights = {}

local CURRENT = {
  winid = -1,
  bufnr = -1,
  tabnr = -1,
}
local OTHER = {
  vimade_fade_active = false,
  vimade_focus_active = false,
  is_dark = false,
}
local DEFAULTS = TYPE.extend(DEFAULT_RECIPE.Default(), {
  basebg = nil,
  ncmode = 'buffers',
  -- disabled for absolute peformance
  -- can be enabled to re-check highlights each tick
  nohlcheck = true,
  -- TODO update this doc
  -- tint must be an object where the values tinting functions ()
  -- object is also accepted and passed to the DEFAULT tinter
  -- tinters are executed in order (from index=0 first), the previous value is passed as highlight
  -- to the next tinter
  --
  -- example:
  -- {
    -- id = function() end -- not required by default your function will use the object key as the id.  However if you want your tint to be recalculated based on an external change the id needs to return a different value (think of it as a cache_key)
    -- interpolate = function (name, highlight, target, optional_win) end
    --   retur
    -- }
  --
  -- 
  --tint = {TINT.DEFAULT},
  --tint = {function(win)
    --return {
      --bg = {
        --rgb = {math.random(0,255), math.random(0,255), math.random(0,255)},
        --intensity = math.random()
      --}
    --}}
  --end,
  fadelevel = 0.4, --or function(win) -> value
  fademinimap = nil,
  groupdiff = true,
  groupscrollbind = false,
  link = {
    -- runs the default_matcher with this config
    -- any value that is a config goes to the
    -- default matcher
    --
    -- this is set to an empty object to enforce the default condition
    -- for legacy conditions
    default = {
      buf_name = nil,
      buf_opts = nil,
      buf_vars = nil,
      win_opts = nil,
      win_vars = nil,
      win_config = nil,
    },
    -- include_alternative matchers if desired
    -- function(win, activeWin) end
  },
  blocklist = {
    -- runs the default_matcher with this config
    -- any value that is a config goes to the
    -- default matcher
    --
    -- this is set to an empty object to enforce the default condition
    -- for legacy conditions
    default = {
      highlights = {
        laststatus_3 = function(win, active)
          -- Global statusline, laststatus=3, is currently disabled as multiple windows take ownership
          -- of the StatusLine highlight (see #85).
          if vim.go.laststatus == 3 then
              -- you can also return tables (e.g. {'StatusLine', 'StatusLineNC'})
              return 'StatusLine'
          end
        end,
        -- Prevent ActiveTabs from highlighting.
        'TabLineSel',
        -- Pmenu should only affect menus like the cmdline autocomplete, which are linked to the active window.
        -- When using Focus mode, these need to be disabled.
        'Pmenu',
        'PmenuSel',
        'PmenuKind',
        'PmenuKindSel',
        'PmenuExtra',
        'PmenuExtraSel',
        'PmenuSbar',
        'PmenuThumb',
        -- Lua patterns are supported, just put the text between / symbols:
        -- '/^StatusLine.*/' -- will match any highlight starting with "StatusLine"
      },
      buf_name = nil,
      -- terminal is temporarily disabled until proper fading is added
      buf_opts = {buftype = {'prompt'}},
      buf_vars = nil,
      win_opts = nil,
      win_vars = nil,
      win_type = nil,
      win_config = nil,
    },
    block_inactive_floats = function (win, active)
      return win.win_config.relative ~= '' and
        (win ~= active or win.buf_opts.buftype =='terminal') and true or false
    end,
  },
})

M.defaults = function ()
  return TYPE.deep_copy(DEFAULTS)
end

local check_fields = function (fields, next, current, defaults, return_state)
  local modified = false
  for i, field in pairs(fields) do
    local value = next[field] 
    if value == nil then
      value = defaults[field]
    end
    if current[field] ~= value then
      current[field] = value
      modified = true
    end
  end
  return modified and return_state or M.READY
end

M.setup = function (config)
  local overlay = {}
  local recipe = config and config.recipe
  if recipe and #recipe > 0 then
    if type(recipe[1]) == 'string' then
      local required = {}
      for part in string.gmatch(recipe[1], '[^:]*') do
        table.insert(required, part)
      end
      local file = string.lower(required[1]) 
      local name = (required[2] and required[2] ~= '') and required[2] or file
      if not file and name then
        file = string.lower(name)
      end
      name = string.upper(string.sub(name,1,1)) .. string.lower(string.sub(name, 2))
      local module = require('vimade.recipe.' .. file)
      module = module[name]
      recipe = module(recipe[2] or {})
      overlay = TYPE.extend(overlay, recipe)
    end
  end
  config.recipe = nil
  overlay = TYPE.extend(overlay, TYPE.deep_copy(config))
  local external_overlay = {}
  for i, field in ipairs({'usecursorhold', 'checkinterval', 'enablefocusfading', 'normalid', 'normalncid'}) do
    if config[field] ~= nil then
      external_overlay[field] = config[field]
    end
  end
  M.vimade_lua = overlay
  vim.g.vimade_overlay = external_overlay
end

M.getInfo = function ()
  result = {renderer = 'lua', [vim.type_idx]=vim.types.dictionary}
  for key, value in pairs(M) do
    if type(value) ~= 'function'
      and key ~= 'style'
      and key ~= 'global_ns'
      and key ~= 'global_highlights' then
      result[key] = value
    end
  end
  return result
end

M.refresh_global_ns = function ()
  M.global_ns = REAL_NAMESPACE.refresh(0, true)
  if M.global_ns.modified  then
    M.global_highlights = M.global_ns.highlights
  end
end

M.callbacks = {}

M.refresh = function (override_tick_state)
  M.now = vim.loop.now()
  M.tick_id = next_tick_id()
  M.tick_state = override_tick_state or M.READY
  -- no reason to re-copy vimade_lua we aren't going to change it
  local vimade = TYPE.shallow_extend(TYPE.deep_copy(vim.g.vimade), M.vimade_lua)
  local current = {
    winid = tonumber(vim.api.nvim_get_current_win()),
    bufnr = tonumber(vim.api.nvim_get_current_buf()),
    tabnr = tonumber(vim.fn.tabpagenr()),
  }

  if vimade.nohlcheck ~= nil then
    M.nohlcheck = TYPE.num_to_bool(vimade.nohlcheck)
  else
    M.nohlcheck = DEFAULTS.nohlcheck
  end

  M.tick_state = BIT_BOR(M.tick_state, check_fields({
    'normalid',
    'normalncid',
  }, vimade, M, DEFAULTS, M.RECALCULATE))
  M.tick_state = BIT_BOR(M.tick_state, check_fields({
    'is_dark',
    'colorscheme',
    'termguicolors',
  }, {
    is_dark = vim.go.background == 'dark',
    colorscheme = vim.g.colors_name,
    termguicolors = vim.go.termguicolors,
  }, M, OTHER, BIT_BOR(M.RECALCULATE, M.DISCARD_NS, M.CHANGED)))
  M.tick_state = BIT_BOR(M.tick_state, check_fields({
    'vimade_fade_active',
    'vimade_focus_active',
  }, {
    vimade_fade_active = vim.g.vimade_fade_active == 1,
    vimade_focus_active = FOCUS.global_focus_enabled(),
  }, M, OTHER, M.CHANGED))
  M.tick_state = BIT_BOR(M.tick_state, check_fields({
    'ncmode',
  }, vimade, M, DEFAULTS, M.CHANGED))
  M.tick_state = BIT_BOR(M.tick_state, check_fields({
    'winid',
    'bufnr',
    'tabnr',
  }, current, M.current, CURRENT,  M.CHANGED))

  -- will be handled in win_state --
  -- link and blocklist are merged at the name level (user overlay has priority)
  M.link = TYPE.extend({}, DEFAULTS.link, vimade.link)
  M.blocklist = TYPE.extend({}, DEFAULTS.blocklist, vimade.blocklist)

  M.basebg = vimade.basebg ~= '' and vimade.basebg or DEFAULTS.basebg
  M.groupdiff = TYPE.num_to_bool(vimade.groupdiff, DEFAULTS.groupdiff)
  M.groupscrollbind = TYPE.num_to_bool(vimade.groupscrollbind, DEFAULTS.groupscrollbind)
  M.fademinimap = TYPE.num_to_bool(vimade.fademinimap, DEFAULTS.fademinimap)
  M.tint = vimade.tint or DEFAULTS.tint
  M.fadelevel = vimade.fadelevel or DEFAULTS.fadelevel
  M.style = vimade.style or DEFAULTS.style

  -- already checked --
  M.nc_windows = M.ncmode == 'windows' or (M.ncmode == 'focus' and M.vimade_focus_active)
  M.nc_buffers = M.ncmode == 'buffers'
  -- if you don't choose one of the above, everything is highlighted

  if not M.global_ns or not M.nohlcheck
    or BIT_BAND(M.CHANGED, M.tick_state) > 0 then
    M.refresh_global_ns()
    if M.global_ns.modified then
      M.tick_state = BIT_BOR(M.RECALCULATE, M.tick_state)
    end
  end

end

return M
